sql_generators/serp_events/templates/desktop_query.sql

```diff
@@ -0,0 +1,262 @@
+WITH raw_serp_events AS (
+  SELECT
+    *,
+    mozfun.map.get_key(event.extra, 'impression_id') AS impression_id,
+    event
+  FROM
+    `moz-fx-data-shared-prod.firefox_desktop_stable.events_v1`,
+    UNNEST(events) AS event
+  WHERE
+    event.category = 'serp'
+    -- allow events related to an impression_id to span 2 submission dates
+    -- we restrict to event sequences started on a single date below
+    AND DATE(submission_timestamp) >= DATE_SUB(@submission_date, INTERVAL 1 DAY)
+    AND DATE_DIFF(DATE(submission_timestamp), @submission_date, DAY) IN (0, 1)
+),
+serp_event_counts AS (
+  SELECT
+    impression_id,
+    COUNTIF(event.name = 'impression') AS n_impressions,
+    COUNTIF(event.name = 'engagement') AS n_engagements,
+    COUNTIF(event.name = 'abandonment') AS n_abandonments,
+  FROM
+    raw_serp_events
+  GROUP BY
+    impression_id
+),
+filtered_impression_ids AS (
+  -- select serp sessions/impression IDs with the expected combinations of events
+  SELECT
+    impression_id
+  FROM
+    serp_event_counts
+  WHERE
+    n_impressions = 1
+    AND ((n_engagements >= 1 AND n_abandonments = 0) OR (n_engagements = 0 AND n_abandonments = 1))
+),
+serp_events AS (
+  SELECT
+    *
+  FROM
+    raw_serp_events
+  INNER JOIN
+    filtered_impression_ids
+  USING
+    (impression_id)
+),
+impressions AS (
+  -- pull top-level fields from the impression event
+  SELECT
+    impression_id,
+    submission_timestamp,
+    client_info.client_id AS glean_client_id,
+    metrics.uuid.legacy_telemetry_client_id AS legacy_telemetry_client_id,
+    ping_info.seq AS ping_seq,
+    event.timestamp AS event_timestamp,
+    normalized_channel,
+    normalized_country_code,
+    client_info.os,
+    mozfun.norm.browser_version_info(client_info.app_display_version) AS browser_version_info,
+    sample_id,
+    ping_info.experiments,
+    -- SERP session characteristics
+    CAST(mozfun.map.get_key(event.extra, 'is_shopping_page') AS bool) AS is_shopping_page,
+    mozfun.map.get_key(event.extra, 'provider') AS search_engine,
+    mozfun.map.get_key(event.extra, 'source') AS sap_source,
+    CAST(mozfun.map.get_key(event.extra, 'tagged') AS bool) AS is_tagged
+  FROM
+    serp_events
+  WHERE
+    event.name = 'impression'
+    -- restrict to sessions that started on the target submission date
+    AND DATE(submission_timestamp) = @submission_date
```

Comment:
And then this should be  `AND DATE(submission_timestamp) = DATE_SUB(@submission_date, INTERVAL 1 DAY)`?

This CTE determines the date for the final rows that will be written by this run. The query should write rows for the day before the latest date with complete data.